src/hotspot/share/gc/shared/slidingForwarding.cpp

```diff
@@ -0,0 +1,156 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "gc/shared/gc_globals.hpp"
+#include "gc/shared/slidingForwarding.hpp"
+#include "utilities/ostream.hpp"
+#include "utilities/powerOfTwo.hpp"
+
+#ifdef _LP64
+
+HeapWord* const SlidingForwarding::UNUSED_BASE = reinterpret_cast<HeapWord*>(0x1);
+
+SlidingForwarding::SlidingForwarding(MemRegion heap, size_t region_size_words)
+  : _heap_start(heap.start()),
+    _num_regions(((heap.end() - heap.start()) / region_size_words) + 1),
+    _region_size_words_shift(log2i_exact(region_size_words)),
+  _target_base_table(nullptr),
+  _fallback_table(nullptr) {
+  assert(_region_size_words_shift <= NUM_COMPRESSED_BITS, "regions must not be larger than maximum addressing bits allow");
+  size_t heap_size_words = heap.end() - heap.start();
+  if (UseSerialGC && heap_size_words <= (1 << NUM_COMPRESSED_BITS)) {
+    // In this case we can treat the whole heap as a single region and
+    // make the encoding very simple.
+    _num_regions = 1;
+    _region_size_words_shift = log2i_exact(round_up_power_of_2(heap_size_words));
+  }
+}
+
+SlidingForwarding::~SlidingForwarding() {
+  if (_target_base_table != nullptr) {
+    FREE_C_HEAP_ARRAY(HeapWord*, _target_base_table);
+    _target_base_table = nullptr;
+  }
+  if (_fallback_table != nullptr) {
+    delete _fallback_table;
+    _fallback_table = nullptr;
+  }
+}
+
+void SlidingForwarding::begin() {
+  assert(_target_base_table == nullptr, "Should be uninitialized");
+  size_t max = _num_regions * NUM_TARGET_REGIONS;
+  _target_base_table = NEW_C_HEAP_ARRAY(HeapWord*, max, mtGC);
+  for (size_t i = 0; i < max; i++) {
+    _target_base_table[i] = UNUSED_BASE;
+  }
+}
+
+void SlidingForwarding::end() {
+  assert(_target_base_table != nullptr, "Should be initialized");
+  FREE_C_HEAP_ARRAY(HeapWord*, _target_base_table);
+  _target_base_table = nullptr;
+
+  if (_fallback_table != nullptr) {
+    delete _fallback_table;
+    _fallback_table = nullptr;
+  }
+}
+
+void SlidingForwarding::fallback_forward_to(HeapWord* from, HeapWord* to) {
+  if (_fallback_table == nullptr) {
+    _fallback_table = new FallbackTable();
+  }
+  _fallback_table->forward_to(from, to);
+}
+
+HeapWord* SlidingForwarding::fallback_forwardee(HeapWord* from) const {
+  assert(_fallback_table != nullptr, "fallback table must be present");
+  return _fallback_table->forwardee(from);
+}
+
+FallbackTable::FallbackTable() {
+  for (size_t i = 0; i < TABLE_SIZE; i++) {
+    _table[i]._next = nullptr;
+    _table[i]._from = nullptr;
+    _table[i]._to   = nullptr;
+  }
+}
+
+FallbackTable::~FallbackTable() {
+  for (size_t i = 0; i < TABLE_SIZE; i++) {
+    FallbackTableEntry* entry = _table[i]._next;
+    while (entry != nullptr) {
+      FallbackTableEntry* next = entry->_next;
+      FREE_C_HEAP_OBJ(entry);
+      entry = next;
+    }
+  }
+}
+
+size_t FallbackTable::home_index(HeapWord* from) {
```

Comment:
What is the license for this code?

Reply:
I've written that code using publicly available documentation. Do I need legal review to do that? Alternatively I would use the murmur3 mixer which is public domain.

Reply:
I changed that code to use the murmur3 mixer instead. This is PD and should hopefully not cause troubles.

Reply:
Is it used in hotspot already? The pragmatic thing would be to not introduce code with external licenses, whatever that license is, unless you want a ride in a legal slow path. We felt that pain with zHash. 1/10 would not recommend. Of course if you need it we can do it, but I'd think twice before going down that route.

Reply:
It seems to be used in a couple of places already:

grep -R ff51afd7ed558ccd src
src/jdk.jfr/share/classes/jdk/jfr/internal/EventWriterKey.java:        z = (z ^ (z >>> 33)) * 0xff51afd7ed558ccdL;
src/java.base/share/classes/java/util/SplittableRandom.java:        z = (z ^ (z >>> 33)) * 0xff51afd7ed558ccdL; // MurmurHash3 mix constants
src/java.base/share/classes/java/util/concurrent/ThreadLocalRandom.java:        z = (z ^ (z >>> 33)) * 0xff51afd7ed558ccdL;
src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:        z = (z ^ (z >>> 33)) * 0xff51afd7ed558ccdL;
src/hotspot/share/gc/shared/slidingForwarding.cpp:  val *= 0xff51afd7ed558ccdULL;
src/hotspot/share/gc/shenandoah/shenandoahEvacOOMHandler.cpp:  key *= UINT64_C(0xff51afd7ed558ccd);

Reply:
Sounds good then.

Reply:
If you wouldn't mind, I think this one is even better: https://github.com/iwanowww/jdk/blob/ssc.cascading/src/hotspot/share/oops/klass.cpp?plain=1#L309
@iwanowww is using it for faster type checking. Our very own @rose00 is behind this one, so we can definitely use it. It performs very well (~1ns per hash). The hash algo passes BigCrush (as a CBPNRG) and SMhasher (with the right loop to combine the input blocks). It's basically a fantastic hash function, that we are free to use.

Reply:
This is great, and I we're going to need a central hash-code impl for several purposes, soon (i-hash, several GC hash-tables, perhaps more). However, that one looks like it only does 32bit hashing? Is there a 64-bit variant that we can use?

Reply:
Ok that is great stuff! Maybe it'd be useful to move it to a central place with this PR (alternative full GC fwding), because we're going to need it for other purposes, too (other GC tables, i-hash, faster type checking, maybe more?)

Reply:
Yes - great idea. Maybe somewhere in utilities. We might swap to it with ZGC as well when things settle down there.

Reply:
Ok, I pushed a change that uses @rose00's better hashing. I added/changed the 128-bit multiplication to (hopefully) make it portable. Let's see what GHA has to say about this ;-)